airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/thread_based_concurrent_stream.py

```diff
@@ -2,36 +2,142 @@
-from typing import Any, Iterable, Mapping, Optional
+from queue import Queue
+from typing import Any, Dict, Iterable, List, Mapping, Optional
 
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.message import MessageRepository
 from airbyte_cdk.sources.streams.concurrent.abstract_stream import AbstractStream, PrimaryKey
-from airbyte_cdk.sources.streams.concurrent.availability_strategy import StreamAvailability
+from airbyte_cdk.sources.streams.concurrent.availability_strategy import AbstractAvailabilityStrategy, StreamAvailability
+from airbyte_cdk.sources.streams.concurrent.error_message_parser import ErrorMessageParser
+from airbyte_cdk.sources.streams.concurrent.partition_enqueuer import PartitionEnqueuer
+from airbyte_cdk.sources.streams.concurrent.partition_reader import PartitionReader
+from airbyte_cdk.sources.streams.concurrent.partitions.partition import Partition
+from airbyte_cdk.sources.streams.concurrent.partitions.partition_generator import PartitionGenerator
 from airbyte_cdk.sources.streams.concurrent.partitions.record import Record
+from airbyte_cdk.sources.streams.concurrent.partitions.types import PARTITIONS_GENERATED_SENTINEL, PartitionCompleteSentinel, QueueItem
+from airbyte_cdk.sources.utils.slice_logger import SliceLogger
 
 
 class ThreadBasedConcurrentStream(AbstractStream):
+
+    DEFAULT_TIMEOUT_SECONDS = 300
+
+    def __init__(
+        self,
+        partition_generator: PartitionGenerator,
+        max_workers: int,
+        name: str,
+        json_schema: Mapping[str, Any],
+        availability_strategy: AbstractAvailabilityStrategy,
+        primary_key: Optional[PrimaryKey],
+        cursor_field: Optional[str],
+        error_display_message_parser: ErrorMessageParser,
+        slice_logger: SliceLogger,
+        message_repository: MessageRepository,
+        timeout_seconds: int = DEFAULT_TIMEOUT_SECONDS,
+    ):
+        self._stream_partition_generator = partition_generator
+        self._max_workers = max_workers
+        self._threadpool = concurrent.futures.ThreadPoolExecutor(max_workers=self._max_workers, thread_name_prefix="workerpool")
+        self._name = name
+        self._json_schema = json_schema
+        self._availability_strategy = availability_strategy
+        self._primary_key = primary_key
+        self._cursor_field = cursor_field
+        self._error_message_parser = error_display_message_parser
+        self._slice_logger = slice_logger
+        self._message_repository = message_repository
+        self._timeout_seconds = timeout_seconds
+
     def read(self) -> Iterable[Record]:
-        raise NotImplementedError
+        """
+        Read all data from the stream (only full-refresh is supported at the moment)
+
+        Algorithm:
+        1. Submit a future to generate the stream's partition to process.
+          - This has to be done asynchronously because we sometimes need to submit requests to the API to generate all partitions (eg for substreams).
```

Comment:
@clnoll's comment:
> Is the idea that this might be IO bound?
Yes, partition generation is IO bound when the partitions depend on records that must be read from a parent stream